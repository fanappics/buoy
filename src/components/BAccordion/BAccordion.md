```js
<div>
  <b-accordion
    label="Accordion 1"
  >
    <p>Whatever content you want to put in</p>
  </b-accordion>
  <b-accordion
    label="Accordion 2"
  >
    <article>
      <h3>Like an article</h3>
      <p>That has all kinds of info in it.</p>
    </article>
  </b-accordion>
</div>
```