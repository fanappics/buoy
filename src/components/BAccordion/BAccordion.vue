<template>
  <section class="b-accordion">
    <div role="heading" class="b-flex b-flex-center">
      <button
        :id="id"
        class="b-accordion-header"
        :aria-expanded="expanded ? 'true' : 'false'"
        :aria-controls="`${id}-content`"
        type="button"
        tabindex="0"
        @click="toggleExpansion"
        @keydown.enter.prevent
        @keyup.enter.prevent="toggleExpansion"
        @keydown.space.prevent
        @keyup.space.prevent="toggleExpansion"
      >
        {{ label }}
        <span :class="{ 'b-arrow': true, 'ion-arrow-down-b': expanded, 'ion-arrow-up-b': !expanded }" />
      </button>
    </div>
    <transition name="b-fade">
      <div
        v-show="expanded"
        :id="`${id}-content`"
        class="b-accordion-content b-flex b-flex-center"
        role="region"
        :aria-labelledby="id"
      >
        <!-- @slot Use for accordion content -->
        <slot name="content"/>
      </div>
    </transition>
  </section>
</template>

<script>
import events from '../../event-bus'

export default {
  name: 'b-accordion',
  props: {
    // Required props
    /**
    * Label text.  Will be used in display label for accordion heading
    */
    label: {
      type: String,
      required: true
    },
    // Optional props
    /**
    * Initializes accordion collapsed when true.
    */
    collapsed: Boolean,
    id: {
      type: String,
      default () {
        return `accordion-${this._uid}`
      }
    }
  },

  created () {
    events.$on(`collapse-${this.id}`, this.collapse)
    events.$on('collapse', this.collapse)
    events.$on('expand', this.expand)
    events.$on(`expand-${this.id}`, this.expand)
  },

  beforeDestroy () {
    events.$off(`collapse-${this.id}`, this.collapse)
    events.$off('collapse', this.collapse)
    events.$off('expand', this.expand)
    events.$off(`expand-${this.id}`, this.expand)
  },

  data: function() {
    return {
      expanded: !this.collapsed
    }
  },

  methods: {
    toggleExpansion () {
      this.expanded = !this.expanded
    },
    collapse () {
      this.expanded = false
    },
    expand () {
      this.expanded = true
    }
  },

  inject: {
    parentScope: {
      default: null
    }
  }
}
</script>

<style lang="stylus" scoped>
  @import "../../styles/colors"

  .b-accordion
    display flex
    flex 1
    flex-direction column
    margin-bottom 1rem

  button.b-accordion-header
    background-color $accordion-background
    border none
    border-bottom 1px solid $accordion-border
    border-radius 0px
    flex-grow 1
    font-size 1rem
    font-weight 600
    line-height 1.5rem
    padding 0
    padding-bottom 0.5625rem
    text-align left
    &:active, &:focus, &:focus-within
      border-color $input-focus
      color $input-focus
      outline none
  
  .b-arrow
    float right
    font-size 1.5rem

  .b-accordion-content
    flex-grow 1
    padding-top 1.25rem

  .b-fade-enter-active, .b-fade-leave-active
    transition opacity .3s

  .b-fade-enter, .b-fade-leave-to
    opacity 0

</style>
