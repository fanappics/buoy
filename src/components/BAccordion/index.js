import BAccordion from './BAccordion.vue'

/* istanbul ignore next */
BAccordion.install = function install (Vue) {
  Vue.component(BAccordion.name, BAccordion)
}

export default BAccordion
