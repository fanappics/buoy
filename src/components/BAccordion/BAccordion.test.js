import { mount, shallow } from 'vue-test-utils'
import BAccordion from './BAccordion'

describe('BAccordion', () => {
  describe('basic accordion', () => {
    it('mounted html should match snapshot', () => {
      const wrapper = shallow(BAccordion, {
        propsData: {
          id: 'accordion-1',
          label: 'hello world'
        }
      })
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('collapsed accordion', () => {
    it('mounted html should match snapshot', () => {
      const wrapper = shallow(BAccordion, {
        propsData: {
          id: 'accordion-1',
          label: 'hello world',
          collapsed: true
        }
      })
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('header actions', () => {
    it('triggers toggleExpansion when header button is clicked', () => {
      const spy = jest.fn()
      const wrapper = mount(BAccordion,
        {
          propsData: {
            id: 'accordion-1',
            label: 'hello world'
          },
          methods: { toggleExpansion: spy }
        }
      )
      wrapper.find('section > div > button').trigger('click')
      expect(spy).toHaveBeenCalled()
    })
    it('triggers toggleExpansion when focus is on header and enter is pressed', () => {
      const spy = jest.fn()
      const wrapper = mount(BAccordion,
        {
          propsData: {
            id: 'accordion-1',
            label: 'hello world'
          },
          methods: { toggleExpansion: spy }
        }
      )
      wrapper.find('section > div > button').trigger('keyup.enter')
      expect(spy).toHaveBeenCalled()
    })
    it('triggers toggleExpansion when focus is on header and spacebar is pressed', () => {
      const spy = jest.fn()
      const wrapper = mount(BAccordion,
        {
          propsData: {
            id: 'accordion-1',
            label: 'hello world'
          },
          methods: { toggleExpansion: spy }
        }
      )
      wrapper.find('section > div > button').trigger('keyup.space')
      expect(spy).toHaveBeenCalled()
    })
  })
})
