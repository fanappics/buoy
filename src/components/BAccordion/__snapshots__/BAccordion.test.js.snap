// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BAccordion basic accordion mounted html should match snapshot 1`] = `
<section class="b-accordion">
    <div role="heading" class="b-flex b-flex-center">
        <button id="accordion-1" aria-expanded="true" aria-controls="accordion-1-content" type="button" tabindex="0" class="b-accordion-header">
            hello world
            <span class="b-arrow ion-arrow-down-b"></span></button>
    </div>
    <div id="accordion-1-content" role="region" aria-labelledby="accordion-1" class="b-accordion-content b-flex b-flex-center" name="b-fade"></div>
</section>
`;

exports[`BAccordion collapsed accordion mounted html should match snapshot 1`] = `
<section class="b-accordion">
    <div role="heading" class="b-flex b-flex-center">
        <button id="accordion-1" aria-expanded="false" aria-controls="accordion-1-content" type="button" tabindex="0" class="b-accordion-header">
            hello world
            <span class="b-arrow ion-arrow-up-b"></span></button>
    </div>
    <div id="accordion-1-content" role="region" aria-labelledby="accordion-1" class="b-accordion-content b-flex b-flex-center" style="display: none;" name="b-fade"></div>
</section>
`;
